In [61]:
import pandas as pd
import numpy as np
from matplotlib import pyplot as plt

In [66]:
df=pd.read_csv('bengaluru_house_prices.csv')
df.head()

area_type   availability                  location       size  \
0  Super built-up  Area         19-Dec  Electronic City Phase II      2 BHK   
1            Plot  Area  Ready To Move          Chikka Tirupathi  4 Bedroom   
2        Built-up  Area  Ready To Move               Uttarahalli      3 BHK   
3  Super built-up  Area  Ready To Move        Lingadheeranahalli      3 BHK   
4  Super built-up  Area  Ready To Move                  Kothanur      2 BHK   

   society total_sqft  bath  balcony   price  
0  Coomee        1056   2.0      1.0   39.07  
1  Theanmp       2600   5.0      3.0  120.00  
2      NaN       1440   2.0      3.0   62.00  
3  Soiewre       1521   3.0      1.0   95.00  
4      NaN       1200   2.0      1.0   51.00

# Which columns have NaN values

In [16]:
df.isna().any()

area_type       False
availability    False
location         True
size             True
society          True
total_sqft      False
bath             True
balcony          True
price           False
dtype: bool

# How many rows and columns along with non-null values

In [17]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 13320 entries, 0 to 13319
Data columns (total 9 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   area_type     13320 non-null  object 
 1   availability  13320 non-null  object 
 2   location      13319 non-null  object 
 3   size          13304 non-null  object 
 4   society       7818 non-null   object 
 5   total_sqft    13320 non-null  object 
 6   bath          13247 non-null  float64
 7   balcony       12711 non-null  float64
 8   price         13320 non-null  float64
dtypes: float64(3), object(6)
memory usage: 936.7+ KB


In [18]:
df.shape

(13320, 9)

In [27]:
df.groupby('area_type').aggregate('count')

availability  location  size  society  total_sqft  bath  \
area_type                                                                       
Built-up  Area                2418      2418  2418     1215        2418  2410   
Carpet  Area                    87        87    87       54          87    87   
Plot  Area                    2025      2025  2009      311        2025  2009   
Super built-up  Area          8790      8789  8790     6238        8790  8741   

                      balcony  price  
area_type                             
Built-up  Area           2310   2418  
Carpet  Area               82     87  
Plot  Area               1837   2025  
Super built-up  Area     8482   8790

In [29]:
df.groupby('area_type')['area_type'].aggregate('count')

area_type
Built-up  Area          2418
Carpet  Area              87
Plot  Area              2025
Super built-up  Area    8790
Name: area_type, dtype: int64

# Sum of Nan in each columns

In [33]:
df.isnull().sum()

area_type          0
availability       0
location           1
size              16
society         5502
total_sqft         0
bath              73
balcony          609
price              0
dtype: int64

In [34]:
df.isna().sum()

area_type          0
availability       0
location           1
size              16
society         5502
total_sqft         0
bath              73
balcony          609
price              0
dtype: int64

In [67]:
df=df.drop(['society','availability','area_type'],axis='columns')
df

location       size total_sqft  bath  balcony   price
0      Electronic City Phase II      2 BHK       1056   2.0      1.0   39.07
1              Chikka Tirupathi  4 Bedroom       2600   5.0      3.0  120.00
2                   Uttarahalli      3 BHK       1440   2.0      3.0   62.00
3            Lingadheeranahalli      3 BHK       1521   3.0      1.0   95.00
4                      Kothanur      2 BHK       1200   2.0      1.0   51.00
...                         ...        ...        ...   ...      ...     ...
13315                Whitefield  5 Bedroom       3453   4.0      0.0  231.00
13316             Richards Town      4 BHK       3600   5.0      NaN  400.00
13317     Raja Rajeshwari Nagar      2 BHK       1141   2.0      1.0   60.00
13318           Padmanabhanagar      4 BHK       4689   4.0      1.0  488.00
13319              Doddathoguru      1 BHK        550   1.0      1.0   17.00

[13320 rows x 6 columns]

# We can use str.split() function to split the strings with the delimiter and mention the number of occurence of delimiter and expand=True (when we want the strings in seperate columns or expand=False when we want to get it in list.

In [68]:
new_size=df['size'].str.split(' ',n=1,expand=True)
new_size

0        1
0      2      BHK
1      4  Bedroom
2      3      BHK
3      3      BHK
4      2      BHK
...   ..      ...
13315  5  Bedroom
13316  4      BHK
13317  2      BHK
13318  4      BHK
13319  1      BHK

[13320 rows x 2 columns]

In [69]:
df.drop('size',axis='columns',inplace=True)

location total_sqft  bath  balcony   price
0      Electronic City Phase II       1056   2.0      1.0   39.07
1              Chikka Tirupathi       2600   5.0      3.0  120.00
2                   Uttarahalli       1440   2.0      3.0   62.00
3            Lingadheeranahalli       1521   3.0      1.0   95.00
4                      Kothanur       1200   2.0      1.0   51.00
...                         ...        ...   ...      ...     ...
13315                Whitefield       3453   4.0      0.0  231.00
13316             Richards Town       3600   5.0      NaN  400.00
13317     Raja Rajeshwari Nagar       1141   2.0      1.0   60.00
13318           Padmanabhanagar       4689   4.0      1.0  488.00
13319              Doddathoguru        550   1.0      1.0   17.00

[13320 rows x 5 columns]

# New column 'size' created in the original Dataframe 

In [71]:
df['size']=new_size[0]
df['size']

0        2
1        4
2        3
3        3
4        2
        ..
13315    5
13316    4
13317    2
13318    4
13319    1
Name: size, Length: 13320, dtype: object

In [72]:
df

location size total_sqft  bath  balcony   price
0      Electronic City Phase II    2       1056   2.0      1.0   39.07
1              Chikka Tirupathi    4       2600   5.0      3.0  120.00
2                   Uttarahalli    3       1440   2.0      3.0   62.00
3            Lingadheeranahalli    3       1521   3.0      1.0   95.00
4                      Kothanur    2       1200   2.0      1.0   51.00
...                         ...  ...        ...   ...      ...     ...
13315                Whitefield    5       3453   4.0      0.0  231.00
13316             Richards Town    4       3600   5.0      NaN  400.00
13317     Raja Rajeshwari Nagar    2       1141   2.0      1.0   60.00
13318           Padmanabhanagar    4       4689   4.0      1.0  488.00
13319              Doddathoguru    1        550   1.0      1.0   17.00

[13320 rows x 6 columns]

In [73]:
df.isna().sum()

location        1
size           16
total_sqft      0
bath           73
balcony       609
price           0
dtype: int64

In [75]:
df.shape

(13320, 6)

Bathroom, size and Balcony have only 73,16 and 609 NaN values out of 13320 rows,hence dropping NA values

In [76]:
df=df.dropna()

In [77]:
df.shape

(12710, 6)

In [80]:
df.isna().sum()

location      0
size          0
total_sqft    0
bath          0
balcony       0
price         0
dtype: int64

In [83]:
df['BHK']=df['size']

/var/folders/qg/9qnm2g51701gxcxzzwr1q97c0000gn/T/ipykernel_66809/3151660415.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['BHK']=df['size']


In [85]:
df.drop('size',axis='columns',inplace=True)
df

/var/folders/qg/9qnm2g51701gxcxzzwr1q97c0000gn/T/ipykernel_66809/669035501.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df.drop('size',axis='columns',inplace=True)


location total_sqft  bath  balcony   price BHK
0      Electronic City Phase II       1056   2.0      1.0   39.07   2
1              Chikka Tirupathi       2600   5.0      3.0  120.00   4
2                   Uttarahalli       1440   2.0      3.0   62.00   3
3            Lingadheeranahalli       1521   3.0      1.0   95.00   3
4                      Kothanur       1200   2.0      1.0   51.00   2
...                         ...        ...   ...      ...     ...  ..
13314         Green Glen Layout       1715   3.0      3.0  112.00   3
13315                Whitefield       3453   4.0      0.0  231.00   5
13317     Raja Rajeshwari Nagar       1141   2.0      1.0   60.00   2
13318           Padmanabhanagar       4689   4.0      1.0  488.00   4
13319              Doddathoguru        550   1.0      1.0   17.00   1

[12710 rows x 6 columns]

In [86]:
df.BHK.astype(int)

0        2
1        4
2        3
3        3
4        2
        ..
13314    3
13315    5
13317    2
13318    4
13319    1
Name: BHK, Length: 12710, dtype: int64

In [89]:
df

location total_sqft  bath  balcony   price BHK
0      Electronic City Phase II       1056   2.0      1.0   39.07   2
1              Chikka Tirupathi       2600   5.0      3.0  120.00   4
2                   Uttarahalli       1440   2.0      3.0   62.00   3
3            Lingadheeranahalli       1521   3.0      1.0   95.00   3
4                      Kothanur       1200   2.0      1.0   51.00   2
...                         ...        ...   ...      ...     ...  ..
13314         Green Glen Layout       1715   3.0      3.0  112.00   3
13315                Whitefield       3453   4.0      0.0  231.00   5
13317     Raja Rajeshwari Nagar       1141   2.0      1.0   60.00   2
13318           Padmanabhanagar       4689   4.0      1.0  488.00   4
13319              Doddathoguru        550   1.0      1.0   17.00   1

[12710 rows x 6 columns]

In [90]:
df.location.unique()

array(['Electronic City Phase II', 'Chikka Tirupathi', 'Uttarahalli', ...,
       '12th cross srinivas nagar banshankari 3rd stage',
       'Havanur extension', 'Abshot Layout'], dtype=object)

In [91]:
df.total_sqft.unique()

array(['1056', '2600', '1440', ..., '1133 - 1384', '774', '4689'],
      dtype=object)

In [96]:
df.bath.max() # quite strange, will check other values for this row

40.0

In [93]:
df.balcony.unique()

array([1., 3., 2., 0.])

In [94]:
df.balcony.max()

3.0

# Data cleaning is required